File: robert-server-ws-rest/src/test/java/fr/gouv/stopc/robertserver/ws/test/matchers/Base64Matcher.java
package fr.gouv.stopc.robertserver.ws.test.matchers;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeDiagnosingMatcher;

import java.util.Base64;

/**
 * Creates a matcher that matches if the examined String is Base64 encoded and
 * matches the specified matcher when it is Base64 decoded. For example:
 * assertThat("Um9iZXJ0IFNlcnZlcg==", isBase64Encoded(equalTo("Robert Server"))
 */
public class Base64Matcher extends TypeSafeDiagnosingMatcher<String> {

    public static Base64Matcher isBase64Encoded(Matcher<String> contentMatcher) {
        return new Base64Matcher(contentMatcher);
    }

    public static String toBase64(byte[] content) {
        return Base64.getEncoder().encodeToString(content);
    }

    public static String toBase64(String content) {
        return toBase64(content.getBytes());
    }

    public static String toBase64(String content, int truncateBytes) {
        final var truncatedContent = new String(content.getBytes(), 0, truncateBytes);
        return toBase64(truncatedContent);
    }

    private final Matcher<String> contentMatcher;

    private Base64Matcher(Matcher<String> contentMatcher) {
        this.contentMatcher = contentMatcher;
    }

    @Override
    protected boolean matchesSafely(String value, Description mismatchDescription) {
        final var content = new String(Base64.getDecoder().decode(value));
        mismatchDescription.appendText("a Base64 encoded string that contains ");
        contentMatcher.describeMismatch(content, mismatchDescription);
        return contentMatcher.matches(content);
    }

    @Override
    public void describeTo(Description description) {
        description.appendText("a Base64 encoded String that contains ");
        contentMatcher.describeTo(description);
    }
}

File: robert-server-test/src/main/kotlin/fr/gouv/stopc/robertserver/test/matchers/Base64Matcher.kt
package fr.gouv.stopc.robertserver.test.matchers

import fr.gouv.stopc.robertserver.common.base64Decode
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeDiagnosingMatcher

/**
 * Creates a matcher that matches if the examined [String] is Base64 encoded and matches the specified matcher when it is Base64 decoded.
 *
 * For example:
 * ```
 * assertThat("Um9iZXJ0IFNlcnZlcg==", isBase64Encoded(equalTo("Robert Server"))
 * ```
 */
fun isBase64Encoded(contentMatcher: Matcher<String>): Base64Matcher {
    return Base64Matcher(contentMatcher)
}

class Base64Matcher internal constructor(private val contentMatcher: Matcher<String>) : TypeSafeDiagnosingMatcher<String>() {

    override fun matchesSafely(value: String, mismatchDescription: Description): Boolean {
        val content = value.base64Decode().decodeToString()
        mismatchDescription.appendText("a Base64 encoded string that contains ")
        contentMatcher.describeMismatch(content, mismatchDescription)
        return contentMatcher.matches(content)
    }

    override fun describeTo(description: Description) {
        description.appendText("a Base64 encoded String that contains ")
        contentMatcher.describeTo(description)
    }
}
